### PR TITLE
VAT exempt and “TAX enabled” handling in back end and Orders API

### PR DESCRIPTION
**Pull request by peterrath**
It is about two related problems about VAT handling: 
1. vat exemption cannot be handled completely (in back-end, API), and 
2. tax enabled / disabled status is not fully handled (in back-end)
1. We realized these issues when tried to use WooCommerce EU VAT Number plugin of WooThemes, which cannot calculate the right taxes on the back-end. There is a ticket about this issue: #368338. According to the answer of support team, we would develop an alternative solution using existing filters of WC, but WooCommerce core does not completely support the above mentioned functions.
2. Another related problem that tax enabled / disabled option is not handled right, when use "Calculate Taxes" button in back end. If some tax rates are set, but tax is disabled (tax enabled option is off), taxes are calculated and saved (but not shown), even though "tax enabled" option is off. After re-enabling taxes, the order totals will be shown. So an order that has not got displayed tax, later has displayed tax.

Base of the solution:
In front-end “is_vat_exempt” data controls if VAT has to be calculated or not. WooCommerce EU VAT Number plugin uses this on front-end:
class-wc-cart.php  calculate_totals()
if ( WC()->customer->is_vat_exempt() ) {
    $this->remove_taxes();
}
WC Back-end and Orders API do not handle vat_exempt status at this moment.

Solution:
class-wc-ajax.php calc_line_taxes() - front end
Apply an additional condition for checking if tax is enabled and there is not vat exempt. If this is true, then calculate taxes on front-end, otherwise not.
TAX exempt can be set by a new woocommerce_ajax_calc_line_taxes_is_vat_exempt filter.
(This condition is similar to the if statement of abstract-wc-order.php->calculate_totals() function.)

class-wc-api-orders.php create_order( $data ) - Orders API
At this moment we cannot find a way how to create an order using the Orders API with tax exempt status. An additional “is_vat_exempt” API field can be used for this reason. If this field is set and true then VAT is not calculated.

Thanks in advance.

### CODE REVIEW

**Comment by mikejolley** [comment]
The backend does not need this;

1. Users are not stored with 'vat exempt'
2. You don't need to calc taxes when you calc totals via backend

Happy to pass through something to prevent tax calc via API, but I think it can be done simpler than your PR.

**Comment by mikejolley** [comment]
My suggestion to update this PR:

1. Add check here for wc_tax_enabled and a check on order post meta called 'is_vat_exempt' https://github.com/woothemes/woocommerce/blob/master/includes/abstracts/abstract-wc-order.php#L626 If is_vat_exempt is 'yes', or wc_tax_enabled is false, return and do nothing.
2. Support the setting of is_vat_exempt meta data via orders API.

cc @claudiosmweb

**Comment by peterrath** [comment]
1. Sorry for disagree with you, but this is a real problem for us, there is a lot of situation when an VAT exempt order which was created on front end should be modified and recalculated. E.g. you would like to give the customer a special price by modifying the line item price of the order. Or if a customer ask over phone to add a new product to his order. 
Normally in these cases the tax has to be recalculated by using calculate taxes button. If the shop administrator accidentally press this button in case of a VAT exempt order, then a wrong order will be generated with none zero tax. Applying the simple modification that we suggested in class-wc-ajax.php this problem can be eliminated, end using WooThemes "WooCommerce EU VAT Number plugin" can handle this situation well on back-end. 
Sorry to write it again, I just want to explain the situation when it causes problem for us. It is really important for me to find a solution, because we use back end heavily, and have to apply EU VAT without any problem.

2. The API solution that you have suggested requires that the data be saved as a post meta before this: https://github.com/woothemes/woocommerce/blob/master/includes/abstracts/abstract-wc-order.php#L626 
This requires an additional customization of the API, because these data is not saved now. So this will be a little bit complicated than a single additional if condition.

At this moment WC do not save these data in front-end, just use them in memory variables. If API require to save it, it will be a different solution than in front-end.
The solution that I suggested does not require saving (similarly to front-end), it uses $data parameter of create_order on the fly. I think this or other solution that does not require saving fits better to the existing front-end vat exempt handling.
I suggest to consider this.

Thanks!

**Comment by mikejolley** [comment]
@peterrath I don't think you understood my comment. If we support is_vat_exempt meta, we can disable tax calculations if the button is pushed. Otherwise you cannot track it.

**Comment by peterrath** [comment]
Ok, I understand now. Thanks.

**Comment by peterrath** [comment]
Hi Mike,

I have implemented the solution that you suggested.
The front-end, back-end and API vat_exemption test was successful. Everything is committed.

Comment: This if clause is not indented: https://github.com/peterrath/woocommerce/edit/vat-exempt-and-tax-enabled-handling/includes/class-wc-ajax.php#L1527 (to line 1612), because GitHub shows all the lines are changed inside if clause, but these are not real changes. This way the real changes can be seen better.

Peter

File: includes/abstracts/abstract-wc-order.php [hunk 1]
```diff
@@ -625,6 +625,11 @@ public function calculate_taxes() {
 		$taxes        = array();
 		$tax_based_on = get_option( 'woocommerce_tax_based_on' );
 
+		// If is_vat_exempt is 'yes', or wc_tax_enabled is false, return and do nothing.
+		if ( 'yes' == $this->is_vat_exempt or ! wc_tax_enabled() ) {
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-order.php]: Use === here

File: includes/class-wc-ajax.php [hunk 2]
```diff
@@ -1521,6 +1521,11 @@ public static function calc_line_taxes() {
 		// Action
 		$items = apply_filters( 'woocommerce_ajax_calc_line_taxes', $items, $order_id, $country, $_POST );
 
+		$is_vat_exempt = get_post_meta ( $order_id, '_is_vat_exempt', true );
```
Comment (mikejolley) [inline includes/class-wc-ajax.php]: remove space before bracket

File: includes/api/class-wc-api-orders.php [hunk 3]
```diff
@@ -471,6 +472,11 @@ public function create_order( $data ) {
 				}
 			}
 
+			// set is vat exempt
+			if ( isset( $data['is_vat_exempt'] ) ) {
+				update_post_meta( $order->id, '_is_vat_exempt', $data['is_vat_exempt'] );
```
Comment (mikejolley) [inline includes/api/class-wc-api-orders.php]: Add `? 'yes' : 'no'` to force values. API should accept bool.

File: includes/class-wc-ajax.php [hunk 4]
```diff
@@ -1521,6 +1521,11 @@ public static function calc_line_taxes() {
...
+	
+		// Tax is calculated only if tax is enabled and order is not vat exempted
+		if ( wc_tax_enabled() && $is_vat_exempt != 'yes' ) {
```
Comment (mikejolley) [inline includes/class-wc-ajax.php]: !==

**Comment by mikejolley** [comment]
@peterrath added feedback

**Comment by peterrath** [comment]
1. Everything is done according to comments of Mike.
2. I suppose that get_order API function has to returned a bool as vat_exempt value, so I also changed the following line, including 'yes' ? true : false conversion.
https://github.com/peterrath/woocommerce/blob/vat-exempt-and-tax-enabled-handling/includes/api/class-wc-api-orders.php#L218
3. This if statment is indented:
https://github.com/peterrath/woocommerce/blob/vat-exempt-and-tax-enabled-handling/includes/class-wc-ajax.php#L1527-L1613

**Comment by mikejolley** [comment]
cc @claudiosmweb needs api docs for 2.6 :)

**Pull request merged**